### DISCUSSION TITLE
🚀 Pandora: Major AI Enhancements, Async Transition, and Modular Updates 🎉

### DISCUSSION THREAD

**Discussion by dh4340** [cmu-sei/GHOSTS#472] [Show and tell]
# 🚀 Pandora: Major AI Enhancements, Async Transition, and Modular Updates 🎉  

I'm excited to unveil a series of transformative updates to **Pandora**, taking its capabilities to the next level with a focus on asynchronous frameworks, modular design, and cutting-edge AI integration.  

---

### **1. Transition to an Async Framework**  
We’ve completely overhauled Pandora’s server to operate on an **asynchronous framework**. This shift enables:  
- **Non-blocking operations**, improving performance and responsiveness under heavy workloads.  
- Seamless handling of multiple requests for real-time applications.  

---

### **2. Enhanced Modular Design**  
To future-proof Pandora and simplify development, the codebase has been refactored into **distinct modules**:  
- **Ease of Maintenance**: Clear separation of concerns for more manageable updates and debugging.  
- **Efficient Development**: Modular architecture reduces dependencies and enhances scalability.  

---

#### **Ollama Integration**  
- Advanced content generation capabilities now integrated across **most endpoints**.  
- Delivers dynamic, high-quality responses for text, audio, and video.  

#### **Voice and Video Generation**  
- **Voice Generation API**: Simulates **realistic conversations** based on actor roles or defaults.  
- **Video Generation API**: Creates **short, professional-quality videos** dynamically.  
  - **Endpoints**: `/video/{file_name}` (GET & POST)  

---

### **4. Video Streaming**  
Pandora now supports **dynamic video streaming** alongside traditional file serving. This feature significantly expands its multimedia capabilities.  

---

### **Key Features Added**:
1. **API Documentation Generation**  
   - Automatic **Swagger documentation** available at `/docs`, providing a comprehensive overview of all routes.

2. **Dynamic Content Generation**  
   - Integrated **Faker** for generating realistic random data in responses.  

3. **Refined Imports**  
   - Improved efficiency of modules like `zipstream`, `docx`, `PIL`, and `pptx`.  

4. **Dynamic File Handling**  
   - Removed hard-coded paths for greater flexibility and security.  

---

### **Expanded HTTP and Content Support**
- **New HTTP Methods**: Added support for **PUT** and **DELETE** requests, enabling broader request handling.  
- **Enhanced Response Types**: Supports ZIP, binary, JSON, CSV, DOCX, PPTX, and more with encapsulated logic.  

---

### **Improved Error Handling**  
- Added detailed, **specific error logging** for better debugging and user feedback.

---

## 🎥 **New Feature Highlights**

1. **Voice Generation API**  
   - Realistic and natural voice simulations for any scenario.  
   - Role flexibility: Define actor roles or use defaults.  

2. **Video Generation API**  
   - Generate professional-quality videos dynamically.  
   - Designed for Hopper and Ampere GPUs with optimised resource utilisation.  

---

## 🔧 **Next Steps**

We’re actively working on:
1. Implementing **unit tests** for all API endpoints to ensure robust and reliable functionality.
2. Expanding support for additional file types and multimedia content.

---

💡 **Feedback Welcome!**  
I would love to hear your thoughts on this update! If you have ideas for improvement or feature requests, open an issue or join the discussion.

**Comment by nblair2** [comment]
hi @dh4340 - I started looking at pandora this evening and was able to get it running very quickly. Thanks for this cool tool!

I have 2 questions:
- how can I enable ssl? I tried updating the `nginx.conf` (uncommenting out the designated lines) and setting `nginx_enabled=True` in the `app.config` but it doesn't look like any connections are being accepted on port 443 (despite it being forwarded to the docker proxy)
- do I need a local ollama instance running ? or is the LLM already integrated into the docker container?

**Reply by dh4340** [reply]
Thanks @nblair2 

### **Question 2** 
Ollama will need be running, I suggest in docker as it makes networking easier if everything is contained in docker. 

### **Question 1**
Make sure Nginx is listening on port 443 inside the container:  

```bash
docker exec -it <nginx_container> netstat -tlnp | grep 443
```

If not, check your `nginx.conf` for:  

```nginx
server {
    listen 443 ssl;
    server_name your_domain_or_ip;
    ssl_certificate /etc/nginx/ssl/fullchain.pem;
    ssl_certificate_key /etc/nginx/ssl/privkey.pem;
}
```

Ensure your SSL certs exist (`ls -l /etc/nginx/ssl/`) and restart Nginx:  

```bash
docker restart <nginx_container>
docker logs <nginx_container>
```

Also, confirm port 443 is open on the host:  

```bash
sudo ufw allow 443/tcp
sudo iptables -L -n | grep 443
```

Test SSL with:  

```bash
curl -v https://your_domain
```  

If issues persist, check the logs inside the Nginx container.

**Reply by nblair2** [reply]
so what I am thinking I need to do is update the `docker-compose.yml` to include `nginx `and `ollama `containers? I just want to make sure that neither capability is supposed to be included in the `ghosts-pandora` container?

**Reply by nblair2** [reply]
@dh4340 something like this:
https://github.com/cmu-sei/GHOSTS/pull/498

I need to add a lot of stuff to the `nginx.conf` for additional protocols but this proxy configuration should work where the only thing that is exposed on the host is port 80 (which redirects to 443) and 443. These are served by an nginx container using the example certificates (can replace with your own), which then forwards the requests to the pandora container (through a docker network so its not exposed on the host). Last, the pandora container can access an ollama container on the same docker network.
